EmojiMe

max-simon opened the pull request:
UI update and better user interaction :)

Commit:
EmojiMe

Better UI

Commit:
Implemented Rotation for Smileys

Smileys werden nun nach Augenstellung gedreht.

max-simon commented:
Das Decoding der anderen Parameter klappt leider gar nicht. Ich wollte den einen Decoder nur für das Bild anwenden (siehe endpointByAdding ...), aber Leider codiert er die anderen Parameter trotzdem nicht (parameterEncoding: .URL). Hab es jetzt zu Fuß gemacht, hast Du noch eine bessere Idee?

`// eigener Endpoint, da Authentifikation und Bild
let endPointWithAuthentification = {(target: MicrosoftFaces) -> Endpoint<MicrosoftFaces> in

```
// TODO: Fix this!
// Irgendwie so ähnlich wie unten, aber .URL funzt nicht :/
var addString = ""
if let unwrappedParameters = target.parameters where unwrappedParameters.count > 0 {
    addString = "?"
    var wasFirst = true
    for (key, value) in unwrappedParameters {
        if !wasFirst {
            addString += "&"
        } else {
            wasFirst = false
        }
        if key != "body" {
            addString += key + "=" + (value as! String)
        }
    }
    addString = String(addString.characters.dropLast(1))

}


let url = target.baseURL.URLByAppendingPathComponent(target.path).absoluteString + addString
let parameterEncodingPic = Moya.ParameterEncoding.Custom({ requestConvertible, parameters in
    let urlRequest = requestConvertible.URLRequest.mutableCopy() as! NSMutableURLRequest

    if let body = parameters?["body"] as? NSData {
        urlRequest.HTTPBody = body
    }

    return (urlRequest, nil)
})

var endpoint: Endpoint<MicrosoftFaces>

if let pic = target.parameters?["body"] {
    var newParameters = target.parameters
    newParameters?.removeValueForKey("body")

    endpoint = Endpoint<MicrosoftFaces>(
        URL: url,
        sampleResponseClosure: { .NetworkResponse(200, target.sampleData) },
        method: target.method,
        parameters: newParameters,
        parameterEncoding: .URL
    )


    endpoint = endpoint.endpointByAdding(parameters: ["body": pic], parameterEncoding: parameterEncodingPic)

} else {
    endpoint = Endpoint<MicrosoftFaces>(
        URL: url,
        sampleResponseClosure: { .NetworkResponse(200, target.sampleData) },
        method: target.method,
        parameters: target.parameters,
        parameterEncoding: .URL
    )
}


return endpoint.endpointByAddingHTTPHeaderFields([
    "Ocp-Apim-Subscription-Key": target.authKey,
    "Content-Type": "application/octet-stream",
    ])
```

}
`

Commit:
Bug fix

API-Abfragen haben unterschiedliche Werte für Gesichtsrechtecke
zurückgegeben -> viele Gesichter nicht erkannt.

nilsvu commented:
Hey Max, so klappts:

``` swift
    let parameterEncodingPic = Moya.ParameterEncoding.Custom({ requestConvertible, parameters in
        let urlRequest = requestConvertible.URLRequest.mutableCopy() as! NSMutableURLRequest
        guard var parameters = parameters else { return (urlRequest, nil) }

        if let body = parameters["body"] as? NSData {
            urlRequest.HTTPBody = body
            parameters.removeValueForKey("body")
        }

        let urlEncodedParameters = parameters.flatMap { key, value in
            return key + "=" + (value as! String) // FIXME: properly convert to String!
            }.joinWithSeparator("&")
        urlRequest.URL = NSURL(string: urlRequest.URL!.absoluteString + (urlEncodedParameters.isEmpty ? "" : "?") + urlEncodedParameters)

        return (urlRequest, nil)
    })
```

Noch eine schönere Variante, die `Rectangle`s zu encoden:

``` swift
            return ["body": imageData, "faceRectangles": faceRectangles.map({ $0.description }).joinWithSeparator(";") ]
```

nilsvu commented:
Btw. so sollte es im Idealfall klappen, wird dran gearbeitet, kannst ja was zu beisteuern ;) https://github.com/Moya/Moya/issues/114

Fix mal bitte noch grad das Repository, also was ich in der Email schrieb und dass du nicht zwei Xcode Projekte im Repo hast. Außerdem die Tests löschen, wenn die leer sind.

Commit:
Code improvements

Thx

max-simon commented:
Vielen Dank, scheint ja echt für alles ne Funktion zu geben 👀

Commit:
Single Face manipulation

Change smiley on single faces by swiping, added 10 new Smileys (only
available on Swiping)

Commit:
Added Comments

nilsvu commented:
Sehr schönes Beispiel, wie wir auf höchst komplexe Technologie anderer Entwickler zurückgreifen und darauf aufbauen können 👏 5/2 Pt. 💯
